### PR TITLE
HDDS-5246. Wait for ever to obtain CA list which is needed during OM/DN startup

### PR DESCRIPTION
**Pull request by bharatviswa504**
## What changes were proposed in this pull request?

Wait forever to obtain the CA list.
And also made retry wait configurable.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5246

## How was this patch tested?

Tested with docker-compose ozonesecure-ha (with few changes to remove WAITFOR scm3.org)

```
docker-compose up scm1.org
docker-compose om1
docker-compose datanode1
```

```
om1_1        | 2021-05-19 09:02:16,082 [main] INFO utils.RetriableTask: Execution of task getCAList failed, will be retried in 10000 ms
om1_1        | 2021-05-19 09:02:26,102 [main] INFO utils.HAUtils: Expected CA list size 4, where as received CA List size 2.
om1_1        | 2021-05-19 09:02:26,103 [main] INFO utils.RetriableTask: Execution of task getCAList failed, will be retried in 10000 ms
om1_1        | 2021-05-19 09:02:36,147 [main] INFO utils.HAUtils: Expected CA list size 4, where as received CA List size 2.
om1_1        | 2021-05-19 09:02:36,147 [main] INFO utils.RetriableTask: Execution of task getCAList failed, will be retried in 10000 ms
om1_1        | 2021-05-19 09:02:46,198 [main] INFO utils.HAUtils: Expected CA list size 4, where as received CA List size 2.
om1_1        | 2021-05-19 09:02:46,198 [main] INFO utils.RetriableTask: Execution of task getCAList failed, will be retried in 10000 ms
om1_1        | 2021-05-19 09:02:56,222 [main] INFO utils.HAUtils: Expected CA list size 4, where as received CA List size 2.
om1_1        | 2021-05-19 09:02:56,222 [main] INFO utils.RetriableTask: Execution of task getCAList failed, will be retried in 10000 ms
om1_1        | 2021-05-19 09:03:06,245 [main] INFO utils.HAUtils: Expected CA list size 4, where as received CA List size 2.
om1_1        | 2021-05-19 09:03:06,245 [main] INFO utils.RetriableTask: Execution of task getCAList failed, will be retried in 10000 ms
om1_1        | 2021-05-19 09:03:16,263 [main] INFO utils.HAUtils: Expected CA list size 4, where as received CA List size 2.
om1_1        | 2021-05-19 09:03:16,263 [main] INFO utils.RetriableTask: Execution of task getCAList failed, will be retried in 10000 ms
om1_1        | 2021-05-19 09:03:26,296 [main] INFO utils.HAUtils: Expected CA list size 4, where as received CA List size 2.
om1_1        | 2021-05-19 09:03:26,296 [main] INFO utils.RetriableTask: Execution of task getCAList failed, will be retried in 10000 ms
om1_1        | 2021-05-19 09:03:36,333 [main] INFO utils.HAUtils: Expected CA list size 4, where as received CA List size 3.
om1_1        | 2021-05-19 09:03:36,333 [main] INFO utils.RetriableTask: Execution of task getCAList failed, will be retried in 10000 ms
om1_1        | 2021-05-19 09:03:46,376 [main] INFO utils.HAUtils: Expected CA list size 4, where as received CA List size 3.
om1_1        | 2021-05-19 09:03:46,376 [main] INFO utils.RetriableTask: Execution of task getCAList failed, will be retried in 10000 ms
```

```
datanode1_1  | 2021-05-19 09:02:50,155 [main] INFO utils.RetriableTask: Execution of task getCAList failed, will be retried in 10000 ms
datanode1_1  | 2021-05-19 09:03:00,179 [main] INFO utils.HAUtils: Expected CA list size 4, where as received CA List size 2.
datanode1_1  | 2021-05-19 09:03:00,179 [main] INFO utils.RetriableTask: Execution of task getCAList failed, will be retried in 10000 ms
datanode1_1  | 2021-05-19 09:03:10,204 [main] INFO utils.HAUtils: Expected CA list size 4, where as received CA List size 2.
datanode1_1  | 2021-05-19 09:03:10,204 [main] INFO utils.RetriableTask: Execution of task getCAList failed, will be retried in 10000 ms
datanode1_1  | 2021-05-19 09:03:20,225 [main] INFO utils.HAUtils: Expected CA list size 4, where as received CA List size 2.
datanode1_1  | 2021-05-19 09:03:20,225 [main] INFO utils.RetriableTask: Execution of task getCAList failed, will be retried in 10000 ms
datanode1_1  | 2021-05-19 09:03:30,266 [main] INFO utils.HAUtils: Expected CA list size 4, where as received CA List size 2.
datanode1_1  | 2021-05-19 09:03:30,266 [main] INFO utils.RetriableTask: Execution of task getCAList failed, will be retried in 10000 ms
datanode1_1  | 2021-05-19 09:03:40,293 [main] INFO utils.HAUtils: Expected CA list size 4, where as received CA List size 3.
datanode1_1  | 2021-05-19 09:03:40,293 [main] INFO utils.RetriableTask: Execution of task getCAList failed, will be retried in 10000 ms
datanode1_1  | 2021-05-19 09:03:50,323 [main] INFO utils.HAUtils: Expected CA list size 4, where as received CA List size 3.
datanode1_1  | 2021-05-19 09:03:50,323 [main] INFO utils.RetriableTask: Execution of task getCAList failed, will be retried in 10000 ms
```

### CODE REVIEW

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/HAUtils.java [hunk 1]
```diff
@@ -413,47 +418,44 @@ public static File getMetaDir(DBDefinition definition,
     return caCertPemList;
   }
 
+  /**
+   * Retry for ever until CA list matches expected count.
+   * @param task - task to get CA list.
+   * @return CA list.
+   * @throws IOException
+   */
+  private static List<String> getCAListWithRetry(Callable task,
```
Comment (adoroszlai) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/HAUtils.java:427]:
```suggestion
  private static List<String> getCAListWithRetry(Callable<List<String>> task,
```

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/HAUtils.java [hunk 2]
```diff
@@ -413,47 +418,44 @@ public static File getMetaDir(DBDefinition definition,
...
+      long waitDuration) throws IOException {
+    RetryPolicy retryPolicy = RetryPolicies.retryForeverWithFixedSleep(
+        waitDuration, TimeUnit.SECONDS);
+    RetriableTask<List<String>> retriableTask =
+        new RetriableTask(retryPolicy, "getCAList", task);
```
Comment (adoroszlai) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/HAUtils.java:432]:
```suggestion
        new RetriableTask<>(retryPolicy, "getCAList", task);
```

**Review by bshashikant — APPROVED** [review]
looks good

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/HAUtils.java [hunk 3]
```diff
@@ -413,47 +418,44 @@ public static File getMetaDir(DBDefinition definition,
...
-
-  private static void checkCertCount(int certCount, int expectedCount)
-      throws SCMSecurityException{
-    if (certCount != expectedCount) {
-      LOG.error("Unable to obtain CA list for SCM cluster, obtained CA list " +
-              "size is {}, where as expected list size is {}",
-          certCount, expectedCount);
-      throw new SCMSecurityException("Unable to obtain complete CA list");
+    caCertPemList = applyFunction.get();
+    caListUpToDate = caCertPemList.size() == expectedCount ? true : false;
```
Comment (adoroszlai) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/HAUtils.java:451]:
```suggestion
    caListUpToDate = caCertPemList.size() == expectedCount;
```

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/HAUtils.java [hunk 4]
```diff
@@ -413,47 +418,44 @@ public static File getMetaDir(DBDefinition definition,
...
-
-
-  private static void checkCertCount(int certCount, int expectedCount)
-      throws SCMSecurityException{
-    if (certCount != expectedCount) {
-      LOG.error("Unable to obtain CA list for SCM cluster, obtained CA list " +
-              "size is {}, where as expected list size is {}",
-          certCount, expectedCount);
-      throw new SCMSecurityException("Unable to obtain complete CA list");
+    caCertPemList = applyFunction.get();
```
Comment (adoroszlai) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/HAUtils.java:450]: Join declaration and assignment.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/HAUtils.java [hunk 5]
```diff
@@ -413,47 +418,44 @@ public static File getMetaDir(DBDefinition definition,
...
+
   private static List<String> waitForCACerts(
       final SupplierWithIOException<List<String>> applyFunction,
       int expectedCount) throws IOException {
-    //TODO: make wait time and sleep time configurable if needed.
     // TODO: If SCMs are bootstrapped later, then listCA need to be
     //  refetched if listCA size is less than scm ha config node list size.
     // For now when Client of SCM's are started we compare their node list
     // size and ca list size if it is as expected, we return the ca list.
     boolean caListUpToDate;
```
Comment (adoroszlai) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/HAUtils.java:448]: I think we can inline this variable (i.e. remove the variable and put the condition directly in `if (...)`).

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/HAUtils.java [hunk 6]
```diff
@@ -413,47 +418,44 @@ public static File getMetaDir(DBDefinition definition,
     return caCertPemList;
   }
 
+  /**
+   * Retry for ever until CA list matches expected count.
+   * @param task - task to get CA list.
+   * @return CA list.
+   * @throws IOException
```
Comment (adoroszlai) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/HAUtils.java:425]:
```suggestion
```

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/HAUtils.java [hunk 7]
```diff
@@ -401,10 +407,9 @@ public static File getMetaDir(DBDefinition definition,
         Collection<String> scmNodes = SCMHAUtils.getSCMNodeIds(configuration);
         int expectedCount = scmNodes.size() + 1;
         if (scmNodes.size() > 1) {
-          caCertPemList = waitForCACerts(
+          caCertPemList = getCAListWithRetry(() -> waitForCACerts(
               () -> scmSecurityProtocolClient.listCACertificate(),
```
Comment (adoroszlai) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/HAUtils.java:411]:
```suggestion
              scmSecurityProtocolClient::listCACertificate,
```

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/HAUtils.java [hunk 8]
```diff
@@ -377,9 +383,9 @@ public static File getMetaDir(DBDefinition definition,
           if (caCertPemList != null && caCertPemList.size() == expectedCount) {
             return caCertPemList;
           }
-          caCertPemList = waitForCACerts(() -> certClient.updateCAList(),
-              expectedCount);
-          checkCertCount(caCertPemList.size(), expectedCount);
+          caCertPemList = getCAListWithRetry(() ->
+              waitForCACerts(() -> certClient.updateCAList(), expectedCount),
```
Comment (adoroszlai) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/HAUtils.java:387]:
```suggestion
              waitForCACerts(certClient::updateCAList, expectedCount),
```

File: hadoop-hdds/common/src/main/resources/ozone-default.xml [hunk 9]
```diff
@@ -2841,4 +2841,18 @@
...
 
+
+  <property>
+    <name>ozone.scm.ca.list.retry.wait.duration</name>
+    <tag>OZONE, SCM, OM, DATANODE</tag>
+    <value>10s</value>
+    <description>SCM client wait duration between each retry to get Scm CA
+      list. OM/Datanode obtain CA list during startup, and wait
+      for the CA List size to be matched with SCM node count size plus
+      1.(Additional one certificate is root CA certificate). If the received
```
Comment (adoroszlai) [inline hadoop-hdds/common/src/main/resources/ozone-default.xml:2852]:
```suggestion
      1. (Additional one certificate is root CA certificate). If the received
```

**Review by adoroszlai — CHANGES_REQUESTED** [review]
LGTM, but I'd like to suggest some some trivial code improvements.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/HAUtils.java (hunk 6 above)
Comment (bharatviswa504) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/HAUtils.java:425]: Why is that, it throws exception right?

Comment (adoroszlai) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/HAUtils.java:425]: >  it throws exception right?

Yes, but `@throws` also needs a description, to document when/why the exception is thrown.  The fact that it is thrown is already documented by the `throws` part of the method declaration.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/HAUtils.java (hunk 5 above)
Comment (bharatviswa504) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/HAUtils.java:447]: Done

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/HAUtils.java (hunk 4 above)
Comment (bharatviswa504) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/HAUtils.java]: Done

**Comment by bharatviswa504** [comment]
Thank You @adoroszlai for the thorough review.
Addressed review comments.

**Review by adoroszlai — APPROVED** [review]
Thanks @bharatviswa504 for updating the patch.

File: hadoop-hdds/common/src/main/resources/ozone-default.xml [hunk 10]
```diff
@@ -2841,4 +2841,18 @@
     </description>
   </property>
 
+
+  <property>
+    <name>ozone.scm.ca.list.retry.wait.duration</name>
```
Comment (xiaoyuyao) [inline hadoop-hdds/common/src/main/resources/ozone-default.xml:2846]: NIT: Can we rename ozone.scm.ca.list.retry.wait.duration to ozone.scm.ca.list.retry.interval to be consistent with other retry settings?

Comment (bharatviswa504) [reply hadoop-hdds/common/src/main/resources/ozone-default.xml]: Done

**Review by xiaoyuyao — APPROVED** [review]
+1 pending CI.

**Pull request merged**